mysql-async/src/main/java/com/github/jasync/sql/db/mysql/MySQLConnection.kt

```diff
@@ -246,17 +251,68 @@ class MySQLConnection @JvmOverloads constructor(
         this.serverVersion = parseVersion(message.serverVersion)
         this.serverStatus = message.statusFlags
 
-        this.connectionHandler.write(
-            HandshakeResponseMessage(
-                configuration.username,
-                configuration.charset,
-                message.seed,
-                message.authenticationMethod,
-                database = configuration.database,
-                password = configuration.password,
-                appName = configuration.applicationName
-            )
+        val switchToSsl = when (this.configuration.ssl.mode) {
+            SSLConfiguration.Mode.Disable -> false
+            SSLConfiguration.Mode.Prefer -> message.supportsSSL()
+            SSLConfiguration.Mode.Require,
+            SSLConfiguration.Mode.VerifyCA,
+            SSLConfiguration.Mode.VerifyFull -> {
+                require(message.supportsSSL()) { "SSL is not supported on server" }
+                true
+            }
+        }
+
+        val sslRequest = SSLRequestMessage(setOfNotNull(
+                CapabilityFlag.CLIENT_PLUGIN_AUTH,
+                CapabilityFlag.CLIENT_PROTOCOL_41,
+                CapabilityFlag.CLIENT_TRANSACTIONS,
+                CapabilityFlag.CLIENT_MULTI_RESULTS,
+                CapabilityFlag.CLIENT_SECURE_CONNECTION,
+                CapabilityFlag.CLIENT_SSL.takeIf { switchToSsl },
+                CapabilityFlag.CLIENT_CONNECT_WITH_DB.takeIf { configuration.database != null },
+                CapabilityFlag.CLIENT_CONNECT_ATTRS.takeIf { configuration.applicationName != null }
+        ))
+
+        val handshakeResponse = HandshakeResponseMessage(
+            sslRequest,
+            configuration.username,
+            configuration.charset,
+            message.seed,
+            message.authenticationMethod,
+            database = configuration.database,
+            password = configuration.password,
+            appName = configuration.applicationName
         )
+
+        if (!switchToSsl) {
+            connectionHandler.write(handshakeResponse)
+            return
+        }
+
+        val channelFuture = connectionHandler.write(sslRequest)
+        channelFuture.addListener { sslRequestFuture ->
+            if (!sslRequestFuture.isSuccess) return@addListener
```

Comment:
what will happen in case of failure? maybe add log warning?

Reply:
I've added a comment, but `connectionHandler.write` is in charge of errors in this case.